src/Features/Core/Portable/CodeRefactorings/SyncNamespace/AbstractSyncNamespaceService.ChangeNamespaceCodeAction.cs

```diff
@@ -0,0 +1,526 @@
+            private ImmutableArray<string> NewNamespaceParts
+            {
+                get
+                {
+                    Debug.Assert(_state.TargetNamespace != null);
+
+                    if (_newNamespaceParts.IsDefault)
+                    {
+                        _newNamespaceParts = _state.TargetNamespace.Split(new[] { '.' }).ToImmutableArray();
+                    }
+
+                    Debug.Assert(_newNamespaceParts.Length > 0);
+                    return _newNamespaceParts;
+                }
+            }
+
+            private ImmutableArray<ISymbol> GetDeclaredSymbolsInContainer(
+                SemanticModel semanticModel, 
+                SyntaxNode node, 
+                CancellationToken cancellationToken)
+            {
+                var declarations = _service.GetMemberDeclarationsInContainer(node);
+                var builder = ArrayBuilder<ISymbol>.GetInstance();
+                foreach (var declaration in declarations)
+                {
+                    var symbol = semanticModel.GetDeclaredSymbol(declaration, cancellationToken);
+                    builder.AddIfNotNull(symbol);
+                }
+
+                return builder.ToImmutableAndFree();
+            }
+
+            protected override async Task<IEnumerable<CodeActionOperation>> ComputeOperationsAsync(CancellationToken cancellationToken)
+            {
+                // Here's the entire process for changing namespace:
+                // 1. Change the namespace declaration, fix references and add imports that might be necessary.
+                // 2. Explicitly merge the diff to get a new solution.
+                // 3. Remove added imports that are unnecessary.
+                // 4. Do another explicit diff merge based on last merged solution.
+                //
+                // The reason for doing explicit diff merge twice is so merging after remove unnecessaty imports can be correctly handled.
+
+                var solutionAfterNamespaceChange = _state.Solution;
+                var referenceDocuments = PooledHashSet<DocumentId>.GetInstance();
+
+                foreach (var id in _state.DocumentIds)
+                {
+                    var result = await ChangeNamespaceToMatchFoldersAsync(solutionAfterNamespaceChange, id, cancellationToken).ConfigureAwait(false);
+                    solutionAfterNamespaceChange = result.Item1;
+                    referenceDocuments.AddRange(result.Item2);
+                }
+
+                var solutionAfterFirstMerge = await MergeDiffAsync(_state.Solution, solutionAfterNamespaceChange, cancellationToken).ConfigureAwait(false);
+
+                // After changing documents, we still need to remove unnecessary imports related to our change.
+                // We don't try to remove all imports that might become unnecessary/invalid after the namespace change, 
+                // just ones that fully matche the old or new namespace.
```

Comment:
this comment is unclear.  Are we doing this intentionally because we think it's the right thing to do?  Or are we doing it because getting that to work is too hard, and we accept that this won't be a perfect change?

Note: i'm ok with either, but we should be clear in comments when we're making a decision that may not be correct, but which we feel is 'acceptable enough' for the quality bar for a feature.  Thakns!

Reply:
Yes, it would be hard to get it right , plus the perf issue that would definitely come with any attempt of doing that.

Reply:
done